src/Service/ExternalUserService.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+namespace App\Service;
+
+use App\Entity\Tenant\UserActivationCode;
+use App\Entity\User;
+use App\Entity\UserRoleTenant;
+use App\Enum\UserTypeEnum;
+use App\Exceptions\CodeGenerationException;
+use App\Exceptions\ExternalUserCodeException;
+use App\Repository\UserActivationCodeRepository;
+use App\Repository\UserRoleTenantRepository;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Component\String\Slugger\AsciiSlugger;
+use Symfony\Component\Uid\Ulid;
+
+class ExternalUserService
+{
+    public const EXTERNAL_USER_DEFAULT_NAME = 'EXTERNAL_NOT_SET';
+
+    public function __construct(
+        private readonly UserActivationCodeRepository $activationCodeRepository,
+        private readonly EntityManagerInterface $entityManager,
+        private readonly Security $security,
+        private readonly string $hashSalt,
+        private readonly UserRoleTenantRepository $userRoleTenantRepository,
+    ) {}
+
+    public function generatePersonalIdentifierHash(string $personalIdentifier): string
+    {
+        return hash('sha512', $this->hashSalt.$personalIdentifier);
+    }
+
+    /**
+     * @throws CodeGenerationException
+     */
+    public function refreshCode(UserActivationCode $code): UserActivationCode
+    {
+        $code->setCode($this->generateExternalUserCode());
+        $code->setCodeExpire((new \DateTime())->add(new \DateInterval('P2D')));
```

Comment:
We should not have this hardcoded two different places (Also found in `ExternalUserActivationCodeInputDataTransformer`). Promote it to a public class constant here, and then reference it. 

(Or better, make it and `.env` value)

Reply:
Fixed